documentation/docs/04-hooks.md

```diff
@@ -2,117 +2,86 @@
 
-This function runs on every incoming request. It generates the `context` object that is available to [endpoint handlers](#routing-endpoints) as `request.context` and that is used to derive the [`session`](#hooks-getsession) object available in the browser.
+If unimplemented, defaults to `({ request, render }) => render(request)`.
 
-If unimplemented, context is `{}`.
+To add custom data to the request, which is passed to endpoints, populate the `request.locals` object, as shown below.
 
 ```ts
-type Incoming = {
+type Request<Locals = Record<string, any>> = {
 	method: string;
 	host: string;
 	headers: Headers;
 	path: string;
+	params: Record<string, string>;
 	query: URLSearchParams;
-	body: string | Buffer | ReadOnlyFormData;
+	rawBody: string | ArrayBuffer;
+	body: string | ArrayBuffer | ReadOnlyFormData | any;
+	locals: Locals;
 };
 
-type GetContext<Context = any> = {
-	(incoming: Incoming): Context;
+type Response = {
+	status?: number;
+	headers?: Headers;
+	body?: any;
 };
+
+type Handle<Locals = Record<string, any>> = ({
+	request: Request<Locals>,
+	render: (request: Request<Locals>) => Promise<Response>
+}) => Response | Promise<Response>;
 ```
 
 ```js
-import * as cookie from 'cookie';
-import db from '$lib/db';
+/** @type {import('@sveltejs/kit').Handle} */
+export async function handle({ request, render }) {
+	request.locals.user = await getUserInformation(request.headers.cookie);
 
-/** @type {import('@sveltejs/kit').GetContext} */
-export async function getContext({ headers }) {
-	const cookies = cookie.parse(headers.cookie || '');
+	const response = await render(request);
 
 	return {
-		user: (await db.get_user(cookies.session_id)) || { guest: true }
+		...response,
+		headers: {
+			...response.headers,
+			'x-custom-header': 'potato'
+		}
 	};
 }
 ```
 
 ### getSession
 
-This function takes the [`context`](#hooks-getcontext) object and returns a `session` object that is [accessible on the client](#modules-$app-stores) and therefore must be safe to expose to users. It runs whenever SvelteKit server-renders a page.
+This function takes the `request` object and returns a `session` object that is [accessible on the client](#modules-$app-stores) and therefore must be safe to expose to users. It runs whenever SvelteKit server-renders a page.
 
 If unimplemented, session is `{}`.
 
 ```ts
-type GetSession<Context = any, Session = any> = {
-	({ context }: { context: Context }): Session | Promise<Session>;
+type GetSession<Locals = Record<string, any>, Session = any> = {
+	(request: Request<Locals>): Session | Promise<Session>;
 };
 ```
 
 ```js
 /** @type {import('@sveltejs/kit').GetSession} */
-export function getSession({ context }) {
+export function getSession(request) {
```

Comment:
Are we confident enough about this that we want to forgo the (single) named parameter here? I have no idea what else in particular might arise, but that doesn't seem entirely watertight.

Reply:
I'm reasonably confident about the single parameter — I can't imagine what else we would want to derive `session` from other than `request`. I'm less confident that we want to pass the whole request in rather than just `locals` but I couldn't quite articulate why, so I figured this was probably fine